easycv/apis/export.py

```diff
@@ -160,31 +181,87 @@ def _export_yolox(model, cfg, filename):
 
     if hasattr(cfg, 'export') and (getattr(cfg.export, 'use_jit', False) or
                                    getattr(cfg.export, 'export_blade', False)):
-        device = 'cuda' if torch.cuda.is_available() else 'cpu'
 
+        # only when we use jit or blade, we need to reparameterize_models before export
+        model = reparameterize_models(model)
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
         model = copy.deepcopy(model)
-        model.eval()
-        model.to(device)
 
         end2end = cfg.export.get('end2end', False)
         if LooseVersion(torch.__version__) < LooseVersion('1.7.0') and end2end:
             raise ValueError('`end2end` only support torch1.7.0 and later!')
 
         batch_size = cfg.export.get('batch_size', 1)
+        static_opt = cfg.export.get('static_opt', True)
+        use_blade = getattr(cfg.export, 'export_blade', False)
+        use_trt_efficientnms = cfg.export.get('use_trt_efficientnms', False)
+
+        # assert image scale and assgin input
         img_scale = cfg.get('img_scale', (640, 640))
+
         assert (
             len(img_scale) == 2
         ), 'Export YoloX predictor config contains img_scale must be (int, int) tuple!'
 
         input = 255 * torch.rand((batch_size, 3) + img_scale)
+        print(input.dtype)
+
+        # assert use_trt_efficientnms only happens when static_opt=True
+        if static_opt is not True:
+            assert (
+                use_trt_efficientnms == False
+            ), 'Export YoloX predictor use_trt_efficientnms=True only when use static_opt=True!'
+
+        # ignore DetPostProcess when use_trt_efficientnms
+        preprocess_fn = None
+        postprocess_fn = None
+        if end2end:
+            preprocess_fn = PreProcess(target_size=img_scale, keep_ratio=True)
+            postprocess_fn = DetPostProcess(max_det=100, score_thresh=0.5)
+
+            if use_trt_efficientnms:
+                logging.warning(
+                    'PAI-YOLOX: use_trt_efficientnms=True during export, we drop DetPostProcess, because trt_efficientnms = detection.boxes.postprocess + DetPostProcess!'
+                )
+                postprocess_fn = None
+
+            if use_blade:
+                logging.warning(
+                    'PAI-YOLOX: End2endModelExportWrapper with preprocess_fn can\'t optimize by blade !'
+                )
+                preprocess_fn = None
+
+        # set model use_trt_efficientnms
+        if use_trt_efficientnms:
+            from easycv.toolkit.blade import create_tensorrt_efficientnms
+            if hasattr(model, 'get_nmsboxes_num'):
+                nmsbox_num = int(model.get_nmsboxes_num(img_scale))
+            else:
+                logging.warning(
+                    'PAI-YOLOX: use_trt_efficientnms encounter model has no attr named get_nmsboxes_num, use 8400 as default!'
+                )
+                nmsbox_num = 8400
```

Comment:
why 8400?   add some comment to define this magic num

Reply:
done